examples/larger-examples/datalog/class-hierarchy-analysis.flix

```diff
@@ -0,0 +1,388 @@
+    ///
+    /// The function takes a `Program` as input and returns a set of relations
+    /// representing the class hierarchy analysis results. The relations include
+    /// `Subtype`, `AvailableMethod`, `AvailableField`, and `ConcreteMethod`.
+    ///
+    /// The `Subtype` relation represents the subtyping relationship between classes
+    /// and interfaces. It includes rules for direct subtyping, class implementing
+    /// an interface, class extending another class, and interface extending another
+    /// interface. It also includes transitive subtyping rules.
+    ///
+    /// The `AvailableMethod` relation represents the available methods for a given
+    /// class. It includes methods defined in the class itself and methods inherited
+    /// from its supertypes.
+    ///
+    /// The `AvailableField` relation represents the available fields for a given
+    /// class. It includes fields defined in the class itself and fields inherited
+    /// from its supertypes.
+    ///
+    /// The `ConcreteMethod` relation represents the methods that are implemented
+    /// on a class or by a super class.
+    ///
+    /// Returns a set of relations representing the class hierarchy analysis results.
+    ///
+    def classHierarchyAnalysis(p: Program): #{ Subtype(String, String), AvailableField(String, String), AvailableMethod(String, String), ConcreteMethod(String, String, String) | r } =
+        let cs    = inject p.classes          into Class;
+        let is    = inject p.interfaces       into Interface;
+        let impls = inject p.classImplements  into ClassImplements;
+        let ce    = inject p.classExtends     into ClassExtends;
+        let ies   = inject p.interfaceExtends into InterfaceExtends;
+        let fs    = inject p.fields           into Field;
+        let ms    = inject p.classMethods     into ClassMethod;
+        let ams   = inject p.abstractMethods  into AbstractMethod;
+        let rules = #{
+            Subtype(x, x)       :- Class(x).
+            Subtype(x, x)       :- Interface(x).
+            Subtype(sub, super) :- Class(sub), Interface(super), ClassImplements(sub, super).
+            Subtype(sub, super) :- Class(sub), Class(super), ClassExtends(sub, super).
+            Subtype(sub, super) :- Interface(sub), Interface(super), InterfaceExtends(sub, super).
+            Subtype(sub, super) :- Subtype(sub, a), Subtype(a, super).
+
+            AvailableField(class, field) :- Field(super, field), Subtype(class, super).
+
+            AvailableMethod(class, method) :- Class(class), ClassMethod(super, method), Subtype(class, super).
+            AvailableMethod(class, method) :- Class(class), AbstractMethod(super, method), Subtype(class, super).
+
+            ConcreteMethod(class, method, class) :- Class(class), ClassMethod(class, method).
+            ConcreteMethod(sub, method, source)  :- Class(sub), Subtype(sub, super), ConcreteMethod(super, method, source).
```

Comment:
Proposal:

ConcreteMethod -> InheritedMethod

And we change the last rule:

```
InheritedMethod(sub, method, source)  :- Class(sub), Subtype(sub, super), InheritedMethod(super, method, source), not ClassMethod(sub, method, sub).
```

Will that work? Such that we account for methods that are overriden?